[1단계 - 상품 관리 기능] 히이로(문제웅) 미션 제출합니다

MoonJeWoong opened the pull request:
안녕하세요 영이, 히이로입니다!

지난 번 체스미션에서 마지막에 급하게 마무리하게 되어 많이 아쉬웠는데 이번에는 더 잘 해보고 싶습니다... 😭 
이번 장바구니 미션에서도 잘 부탁드립니다!!

이번 미션을 진행하며 궁금했던 부분은 다음과 같습니다.

- 테스트 코드를 작성하는 과정에서 이번에는 단위 테스트가 아닌 통합 테스트를 작성했습니다. 그래서 controller에 작성한 HTTP Request를 맵핑하는 메서드들에 대해 요청을 보내고 정상적인 Response를 반환하는지를 테스트했습니다. 그런데 해당 과정에서 update, delete 기능을 수행하는 메서드에 대한 테스트를 작성할 때 어려움이 있었습니다.

- update, delete 기능은 요청 시 product의 id 값을 포함한 데이터들을 body에 담아서 요청해야 합니다. 실제 프로그램이 동작할 때는 페이지 렌더링 시 DB에 저장된 모든 Product 정보를 가져와서 보여줍니다. 그리고 선택한 Product 정보를 이용해 update, delete request 메세지를 보냅니다.
- 문제는 테스트 코드를 작성할 때 있었습니다. 요구사항을 구현하고 나니 controller 단에서 update 혹은 delete를 수행할 때 테스트용 더미 Product를 만들고 해당 더미의 id 값을 따로 조회할 수 있는 방법이 없었습니다. 
  - create 메서드는 요구사항 상 생성한 Product의 id값을 반환할 필요가 없었기에 반환되는 response의 body는 비어있습니다. 그렇다고 create 메서드에 생성된 Product의 id 값을 넣는 것은 테스트를 위해 코드를 변경하는 것이라 여겨져 좋지 못한 방법이라고 생각했습니다.
  - 또 고려했던 방법은 admin 메서드에서 렌더링 완료되어 반환되는 HTML 페이지에서 직접 값을 찾아서 검증하는 방법이었습니다. 허나 이 방법은 HTML 파일에 의존하는 방식이었기에 올바른 방법이라고 생각되진 않았습니다.

- 결과적으로 update, delete에 대한 테스트는 테스트를 진행할 때마다 sql 테이블을 drop하고 다시 만들어서 진행하는 방식으로 진행하게 되었습니다. 이렇게 테스트가 어려워진 것은 현재 controller가 RestController가 아니라 일반 HTML페이지도 반환하는 Controller이기 때문이라고 생각했습니다. 이런 경우 테스트를 좀 더 쉽게 진행할 수 있는 방법은 없었을지 여쭤보고 싶습니다!

마지막으로 이번 레벨2에서 스프링을 처음 사용하기 시작해서 미숙한 점이 아직 많다는 점을 말씀드리고 싶었습니다. 학습해야 할 키워드나 레퍼런스가 혹시 있다면 많이 던져주시면 감사할 것 같습니다! 잘 부탁드립니다~ 🙇

MoonJeWoong committed:
docs : 기능 목록 작성

MoonJeWoong committed:
feat : 상품 목록, 관리자 도구 페이지 연동

MoonJeWoong committed:
feat : 상품 도메인 구현

MoonJeWoong committed:
feat : DB 연동 관련 설정 추가 및 product 테이블 설계 완료

MoonJeWoong committed:
feat : create 폼 데이터 전송 기능 구현

MoonJeWoong committed:
feat : 단일 상품 Create 기능 구현

MoonJeWoong committed:
feat : 전체 상품 Read 기능 구현

MoonJeWoong committed:
feat : 등록된 단일 상품 Update 기능 구현

MoonJeWoong committed:
feat : 등록된 단일 상품 Delete 기능 구현

MoonJeWoong committed:
feat : 상품 목록 페이지 모든 상품 Read 기능 구현

MoonJeWoong committed:
feat : 요청 메세지 값의 null, 공백 검증 기능 구현

MoonJeWoong committed:
refactor: controller 분리 및 예외처리 로직 분리

MoonJeWoong committed:
feat: "/"으로 get 요청시 "/products"로 리다이렉트하는 기능 추가

MoonJeWoong committed:
feat: 통합 테스트 코드 작성

MoonJeWoong committed:
feat: AdminController Delete 메서드 인자 검증 기능 구현

MoonJeWoong committed:
refactor: Service 클래스 전체에 @Transactional 어노테이션 적용

MoonJeWoong committed:
style: 미사용 주석 및 코드 삭제

MoonJeWoong committed:
refactor: 도메인 클래스 내 상수화 진행

MoonJeWoong committed:
refactor: 미사용 클래스 제거

MoonJeWoong committed:
feat: 도메인 예외처리 handler 기능 구현

MoonJeWoong committed:
refactor: 기존 예외처리 메서드 반환 타입 변경

choijy1705 commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,57 @@
+        this.cartService = cartService;
+    }
+
+    @GetMapping("")
```
```suggestion
    @GetMapping
```
생략해도 될것 같네요!

choijy1705 commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,57 @@
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+@Controller
```
RestContorller와 그냥 Controller에 대해 이해하고 계신가요?

MoonJeWoong replied:
@RestController는 @Controller와 @ResponseBody가 결합된 어노테이션인 것으로 알고 있습니다. 그래서 @Controller를 사용했을 때는 반환한 값이 기본적으로 viewResolver를 거치며 view로 렌더링 되어 처리되는 반면 @RestController는 @ResponseBody를 사용한 것처럼 반환 값을 응답 메세지의 body에 담아서 반환하는 과정을 기본적으로 수행하게 됩니다.

여기에서 @Controller를 사용한 이유는 미션 요구사항 상 admin 기본 url에 대한 get 요청 시 resources에 존재하는 HTML 파일을 렌더링 해서 반환했어야 했기 때문입니다.

구현하는 과정에서 페이지를 렌더링해 반환하는 기능과 Rest API 기능을 분리해야하는지 고민했으나 이를 분리하기에는 처리하는 기능 규모가 크지 않다고 판단되어 @Controller를 사용하는 방식으로 결정했습니다!

MoonJeWoong replied:
혹시 현업에서는 이런 경우 기능이 적어도 controller와 RestController를 분리해서 사용하는 지 궁금합니다!

choijy1705 replied:
흠 회사마다 다르겠지만 보통 controller는 안쓸거에요
미션에서는 백엔드에서 프론트부분까지 작업하여 view를 랜더링해주는 특수한 경우일것 같네요
보통 실무에서는 json으로 값을 넘겨주기만 하고 실제 랜더링해주는 부분은 프론트엔드쪽에서 작업해줘서요

기능적으로도 확실히 다르기때문에 저는 RestController와 Controller를 구분하는 편입니다
Controller수가 많아진다면 젤 위에 어노테이션만으로 얘가 어떤 기능을 하는 controller인지 파악할 수 있을것 같네요!

choijy1705 commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,57 @@
+
+    @PostMapping("/create")
+    @ResponseStatus(HttpStatus.CREATED)
+    public void create(@RequestBody @Valid ProductRequest productRequest) {
```
ExceptionHandler에서 사용하신것 처럼 ResponseEntity를 사용해보면 어떨까요?

MoonJeWoong replied:
처음에는 ResponseEntity를 사용해서 다음과 같이 코드를 작성했었습니다.

`return ResponseEntity.ok().build();`

이후 페어와 리팩토링 과정을 거치면서 얘기했던 부분은 응답 메세지 body에 담아서 반환할 것이 없는데 ResponseEntity를 사용해야 하는가? 였습니다. 그리고 ResponseStatus를 사용하면 반환타입이 void인 메서드에서도 HTTP 상태를 클라이언트에게 넘길 수 있다는 것을 알게되어 지금의 코드로 수정했습니다.

하지만 영이가 리뷰를 해주신대로 create와 update 기능 수행 후 작업이 완료된 product를 반환하도록 리팩토링을 하려면 ResponseEntity를 사용하도록 변경해야할 것 같습니다.

MoonJeWoong replied:
저는 위와 같이 생각했는데 혹시 영이는 @ResponseStatus의 사용을 지양하고 ResponseEntity를 사용해서 반환하는 것이 더 좋다고 생각하시는 쪽인지, 혹시 그러시다면 어떠한 이유로 그렇게 생각하시는지 여쭤보고 싶습니다!

choijy1705 replied:
어노테이션은 지금과 코드양이 없을때는 눈에 띄지만
코드양이 많아지면 어노테이션은 아무래도 먼저 보지는 않게됩니다.(제 경험상이긴 합니다)
어노테이션 보다는 실제 로직상에서 return 값이 뭔지를 나타내는게 
개인적으로는 더 좋을것 같다는 생각이듭니다!

choijy1705 commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,57 @@
+        cartService.create(productRequest);
+    }
+
+    @PutMapping("/update")
```
Restful에 대해 학습해보면 좋을것 같아요
URI에서는 자원을 http method로 어떤 동작을 하는지 표현을 하도록 하면 좋을것 같아요!

MoonJeWoong replied:
전반적으로 RESTful이란 무엇인지, REST API는 어떻게 설계해야 하는지 간단하게 학습해봤습니다.

- URI는 자원의 정보를 나타내야 한다.
  - resource의 document 이름으로는 단수명사를 사용한다.
  - resource의 컬렉션, 스토어 이름으로는 복수명사를 사용한다.

- resource에 대한 HTTP 메서드 및 동사는 URI 상에 드러나면 안된다.
- URI 중 변하는 부분은 유일한 값으로 대체한다. (id 등 해당 resource를 특정짓는 고유값)

REST의 개념이 많이 추상적인지라 아직 정확하게는 모르지만 위에 명시한 사항들을 지키는 URI로 수정해봤습니다!

choijy1705 replied:
https://www.youtube.com/watch?v=RP_f5dMoHFc&t=535s

restful 좀 더 자세하게 공부하고 싶으면 요거 한번 보는거 추천드립니다
조금 길어요 ㅠ

choijy1705 commented on src/main/java/cart/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler
+    public ResponseEntity<List<ErrorResponse>> handleMethodArgumentNotValidException(MethodArgumentNotValidException exception) {
+        List<ErrorResponse> responses = exception.getBindingResult().getAllErrors()
+                .stream()
+                .map(error -> new ErrorResponse(
+                        ((FieldError) error).getField() + " : " + error.getDefaultMessage(),
+                        LocalDateTime.now()))
+                .collect(Collectors.toList());
+
+        return ResponseEntity.badRequest().body(responses);
+    }
```
옆으로 너무 긴것 같은데 (한줄에 100자) 조금 수정을 해서 더 깔끔하게 보이게 하면 어떨까요?

choijy1705 commented on src/main/java/cart/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+@ControllerAdvice
```
RestControllerAdvice와 비교해서 어떤게 적절할지 고민해봐주세요!

MoonJeWoong replied:
클라이언트의 요청을 처리하는 컨트롤러가 @Controller 어노테이션을 사용하기에 예외처리를 진행하는 GlobalExceptionHandler도 @ControllerAdvice를 사용해야 한다고 생각했던 것 같습니다.

@RestControllerAdvice 또한 @RestController 처럼 내부적으로 @ResponseBody가 붙어 있다는 차이가 있습니다. 현재 ExceptionHandler는 전체적으로 ResponseEntity를 반환하기 때문에 @RestControllerAdvice를 사용하는 것으로 수정했습니다!

choijy1705 commented on src/main/java/cart/dto/response/ErrorResponse.java:
```diff
@@ -0,0 +1,22 @@
+    private final String message;
+    private final LocalDateTime time;
+
+    public ErrorResponse(String message, LocalDateTime time) {
```
시간을 넣는건 제 입장에서는 조금 생소한데
어떤 이유로 넣어주시나요?? 그냥 궁금해서 물어보는겁니다! 🙏

MoonJeWoong replied:
페어와 같이 진행을 하면서 예외 처리에 대한 응답 메세지에는 언제 발생했는지에 대한 시간도 있으면 좋을 것 같다는 얘기가 나왔어서 추가했습니다.

그런데 다시 생각해보니 시간을 기록하는 것은 로그에 추가하고 에러 메세지에는 어떤 예외 처리가 되었는지에 대한 내용만 넣는 것이 맞다는 생각이 드네요... 예외 처리가 되는 시점에 바로 클라이언트에게 전달되는 메세지라서 시간은 불필요한 정보라고 생각이 되었습니다. 해당 부분은 수정했습니다 ㅎㅎ

choijy1705 commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,10 @@
+     price      INT         NOT NULL,
+     created_at  DATETIME    NOT NULL default current_timestamp,
+     PRIMARY KEY (id)
+);
```
뭔가 정렬이 안되어 있는거 같아요😭

choijy1705 commented on src/test/java/cart/ProductIntegrationTest.java:
```diff
@@ -34,4 +34,15 @@ public void getProducts() {
+                .extract();
+
+        assertThat(result.statusCode()).isEqualTo(HttpStatus.OK.value());
+    }
```
요거랑 AdminContorllerTest의 첫번째 테스트랑 동일한거 아닌가요?

MoonJeWoong replied:
앗 맞습니다... ProductIntegrationTest는 기존 프로젝트를 clone할 때부터 있었던 코드였어서 미처 확인하는 것을 빠뜨렸네요 😭 중복되는 테스트 코드는 삭제하였습니다!

choijy1705 commented on src/test/java/cart/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,131 @@
+    @Test
+    @DisplayName("/admin 으로 get 요청을 보내면 ok 상태코드를 반환한다")
+    void adminTest() {
+        RestAssured.given().log().all()
```
무의식적으로 log, all 을 붙이는게 아닐까요?

MoonJeWoong replied:
동의합니다. 학습 테스트 코드에 적혀있던 형식을 참고해서 작성하다보니 미처 신경쓰지 못한 부분같습니다. 불필요한 로그 출력 부분은 삭제했습니다!

choijy1705 commented on src/test/java/cart/controller/AdminControllerTest.java:
```diff
@@ -0,0 +1,131 @@
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .body(productRequest)
+                .when()
+                .post("/admin/create");
```
이 부분은 여기서 테스트하고 싶은게 아닐것 같아요
dao를 통하여 세팅을 해주면 어떨까요?

MoonJeWoong replied:
맞습니다. 이 부분도 페어와 꽤 많은 시간을 고민했었던 부분이었습니다. 그 때 당시 dao 등을 통해 직접 데이터를 생성하고 테스트를 진행하는 것은 통합테스트가 아닌 단위테스트에서 진행해야 할 부분이라고 생각이 되어서 이런 방식으로 구현하게 되었습니다.

하지만 다시 생각해보니 더미데이터를 세팅하는 부분을 Dao를 통해 직접 해줘도 update, delete 기능 자체에 대해 통합테스트하는 과정에는 영향이 없을 것이라 판단되었습니다. 해당 부분은 applicationContext를 통해 빈으로 등록된 Dao를 직접 찾아서 진행해주는 것으로 수정했습니다!

choijy1705 commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,57 @@
+    @PutMapping("/update")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void update(@RequestBody @Valid ProductUpdateRequest productUpdateRequest) {
+        cartService.update(productUpdateRequest);
```
delete 의 경우에는 noContent로 별다른 응답이 없지만
나머지 create, update 의 경우에는 변경한 값에대한 반환이 있으면 좋을것 같아요
product를 생성했다면 완성된 product를 적절히 dto로 감싸 response로 내려줘야 해당 api를 호출한 곳에서 잘 되었는지 확인할 수 있지 않을까요?

물론 api를 호출하다보면 별다른 응답이나 값을 안내려주는 경우도 있는데 개인적으로 매우 불편했고 레거시 코드라는 느낌이 강하게 들었습니다. 
또한 상태코드만으로 해당 api를 신뢰할 수도 없을것 같아요 결국 api도 사람이 만들다보니 실패했는데 성공 status code를 내려줄수도 있고 성공했는데 실패 status code를 내려줄수도 있으니 더블 체크할수 있는 근거도 될것 같네요
질문주신것처럼 테스트도 어렵게 만드는 요인일것 같고요

create, update시에는 적절하게 변경 혹은 생성된 객체를 반환해주도록 하면 어떨까요?

MoonJeWoong replied:
동의합니다. 사실 페어와 미션을 진행하며 굉장히 불편함을 느꼈으나 요구사항 상 create와 update 수행 이후 추가 작업이 진행되지 않기에 별도의 응답을 반환하는 것이 요구사항과 맞지 않다고 판단했습니다.

하지만 굳이 요구사항에서 필요로 하지 않더라도 영이가 말씀해주신 것처럼 api에 대한 신뢰성을 위해서 Product 정보를 담은 응답 메세지를 반환하도록 구현할 수 있을 것 같습니다.

또 제대로 작업이 이루어졌는지 프론트 단에서도 확인할 수 있는 수단으로도 제공할 수 있을 것 같네요.

Create, Update 수행 시 각각 생성된, 수정된 Product 정보를 담은 응답 메세지를 반환하도록 수정했습니다!

choijy1705 reviewed (changes requested):
안녕하세요 히이로!
다시 만나게되었네요👏
질문 주신 내용과 몇 가지 코멘트 남겼습니다!
히이로의 답변도 궁금해서 바로 머지하지는 않겠습니다

MoonJeWoong committed:
refactor: mapping 어노테이션의 불필요한 인자 삭제

MoonJeWoong committed:
refactor: create, update API 반환 타입 수정

세부사항
- create 메서드는 생성된 Product 정보를 반환하도록 수정
- update 메서드는 변경된 Product 정보를 반환하도록 수정

MoonJeWoong committed:
refactor: 전역 예외처리 클래스 어노테이션 @RestControllerAdvice로 변경

MoonJeWoong committed:
style: 사용하지 않는 주석 제거 및 스타일 정리

MoonJeWoong committed:
refactor: 예외처리 응답 메세지에 시간 정보 삭제

MoonJeWoong committed:
refactor: sql 파일 코드 정렬

MoonJeWoong committed:
test: 불필요 및 중복 테스트 코드 삭제, 테스트 파일 명 변경

MoonJeWoong committed:
refactor: product 상태 값 순서 통일

MoonJeWoong committed:
test: admin 통합 테스트 update, delete 부분 수정

세부사항
- update, delete 테스트를 위한 더미 데이터 생성을 직접 dao를 통해
  수행하도록 수정

MoonJeWoong committed:
refactor: 기존 URI를 RESTful하도록 수정

MoonJeWoong committed:
refactor: Product 컨트롤러 내 URI를 RESTful하게 수정

choijy1705 commented on src/main/java/cart/controller/AdminController.java:
```diff
@@ -0,0 +1,68 @@
+    @DeleteMapping("/products/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void delete(@PathVariable @NotNull Long id) {
+        cartService.delete(id);
```
얘도 ResponseEntity를 사용하면 어떨까요?

choijy1705 commented on src/main/java/cart/domain/Product.java:
```diff
@@ -0,0 +1,35 @@
+    private final Price price;
+    private final String imageUrl;
+
+    public Product(Long id, String name, Integer price, String imageUrl) {
```
price가 null인 경우도 있나요? 없다면 int를 사용하는게 좋겠네요
더 좋은방법은 돈과 관련된 값이라면 BigDecimal을 사용하는게 좋습니다

choijy1705 commented on src/test/java/cart/AdminIntegrationTest.java:
```diff
@@ -0,0 +1,126 @@
+import org.springframework.http.MediaType;
+import org.springframework.test.context.jdbc.Sql;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
domain에 대한 테스트도 있어야 하지 않을까요?😭
1단계 내용들을 잊지 말아주세요!

choijy1705 reviewed (approved):
안녕하세요 히이로!
피드백 반영 잘해주셨네요
이번단계는 머지할께요!
다음 단계 진행하면 몇가지 코멘트도 확인하고 반영해주세요!!
